src/hd.rs

```diff
@@ -181,111 +177,106 @@ impl HDWallet {
-            "xpub network mismatch, {} is {} and not {}",
-            xpub,
-            xpub.network,
-            network
-        );
-        Ok(Self::new(
-            xpub.extended_pubkey,
-            network,
-            xpub.script_type,
-            gap_limit,
-            initial_import_size,
-            rescan_policy,
-        ))
-    }
-
-    pub fn from_xpub(
-        xpub: XyzPubKey,
-        network: Network,
-        gap_limit: u32,
-        initial_import_size: u32,
-        rescan_policy: RescanSince,
-    ) -> Result<Vec<Self>> {
-        ensure!(
-            xpub.matches_network(network),
-            "xpub network mismatch, {} is {} and not {}",
-            xpub,
-            xpub.network,
-            network
-        );
-        Ok(vec![
-            // external chain (receive)
-            Self::new(
-                xpub.extended_pubkey
-                    .derive_pub(&*EC, &[ChildNumber::from(0)])?,
-                network,
-                xpub.script_type,
-                gap_limit,
-                initial_import_size,
-                rescan_policy,
-            ),
-            // internal chain (change)
-            Self::new(
-                xpub.extended_pubkey
-                    .derive_pub(&*EC, &[ChildNumber::from(1)])?,
-                network,
-                xpub.script_type,
-                gap_limit,
-                initial_import_size,
-                rescan_policy,
-            ),
-        ])
-    }
-
-    pub fn from_xpubs(
+    pub fn from_config(
+        descriptors: &[(String, RescanSince)],
         xpubs: &[(XyzPubKey, RescanSince)],
         bare_xpubs: &[(XyzPubKey, RescanSince)],
         network: Network,
         gap_limit: u32,
         initial_import_size: u32,
+        rpc: Arc<RpcClient>,
     ) -> Result<Vec<Self>> {
+        // Descriptors
         let mut wallets = vec![];
-        for (xpub, rescan) in xpubs {
-            wallets.append(
-                &mut Self::from_xpub(
-                    xpub.clone(),
+        for (descriptor, rescan) in descriptors {
+            wallets.push(
+                Self::from_descriptor(
+                    descriptor.clone(),
                     network,
                     gap_limit,
                     initial_import_size,
                     *rescan,
+                    rpc.clone(),
                 )
-                .with_context(|| format!("invalid xpub {}", xpub))?,
+                .with_context(|| format!("invalid descriptor {}", descriptor))?
+                .clone(),
             );
         }
-        for (xpub, rescan) in bare_xpubs {
+
+        // Xpubs
+        for (xyz, rescan) in xpubs {
+            // Change and receiving output descriptors
+            for change in 0..2 {
+                let descriptor = match xyz.script_type {
+                    ScriptType::P2pkh => {
+                        format!("pkh({}/{}/*)", xyz.extended_pubkey.to_string(), change)
```

Comment:
Perhaps this logic could be extracted from here, into something like `Descriptor::from(XyzPubKey)` for the bare case and `Descriptor::from((XyzPubKey, usize))` for the external/internal chains case?

Then you could use `descriptor: xyz.into()` and `descriptor: (xyz, change).into()`.